File: W4D3/Chess2/board.rb
require 'byebug'

Dir["./pieces/*.rb"].each {|file| require_relative file }

class Board
  attr_reader :grid
  
  def initialize 
    @grid = Array.new(8){Array.new(8, [])}
    build_board
  end

  def build_board
    (0..7).each do |i|
      (0..7).each do |j|
        pos = [i, j]
        self[pos] = add_piece(pos)
      end
    end    
    set_color
  end

  def set_color
    [6,7].each do |row|
      (0..7).each do |col|
        self[[row,col]].color = "black"
      end
    end
  end

  def add_piece(pos)
    case pos 
      when [0,0], [0,7], [7,0], [7,7]
        Rook.new(grid, pos)
      when [0,4], [7,4]
        King.new(grid, pos)
      when [0,3], [7,3]
        Queen.new(grid, pos)
      when [0,5], [0,2], [7,2], [7,5]
        Bishop.new(grid, pos)
      when [0,1], [0,6], [7,1], [7,6]
        Knight.new(grid, pos)        
      else
        if pos[0] == 1 || pos[0] == 6
          Pawn.new(grid, pos)
        else
          NullPiece.instance
        end
    end
  end

  def move_piece(start_pos, end_pos)    
    raise "There's no piece there" if self[start_pos].is_a? NullPiece
    raise "invalid position" if !valid_pos?(end_pos)

    if valid_move?(start_pos,end_pos)
      self[end_pos] = self[start_pos] 
      self[start_pos] = NullPiece.instance
    else
      raise 'Invalid move, try again'
    end
  end

  def valid_move?(start_pos,end_pos)
    piece = self[start_pos]
    piece.moves.include?(end_pos)  # method taken from module, returns array of valid moves
  end

  def valid_pos?(pos)
    row, col = pos
    return false if !row.between?(0,7) || !col.between?(0,7)
    true
  end

  def [](pos)
    row, col = pos
    grid[row][col] 
  end

  def []=(pos, value)
    row, col = pos
    grid[row][col] = value
  end
 
  def print_board
    (0...grid.length).each do |i|
      (0...grid.length).each do |j|
        pos = [i, j]
        print "#{self[pos]}  |".ljust(6)
      end
      puts
      puts "---------------------------------------------"  
    end
  end
end


b = Board.new

b.print_board
b.move_piece([1,4],[3,4]) # white pawn
b.move_piece([6,3],[4,3]) # black pawn 
b.move_piece([0,6],[2,5])
b.move_piece([0,4],[1,4])

puts
b.print_board
puts

File: W8D4/Reversi/lib/board.js
let Piece = require("./piece");

/**
 * Returns a 2D array (8 by 8) with two black pieces at [3, 4] and [4, 3]
 * and two white pieces at [3, 3] and [4, 4]
 */
function _makeGrid () {
  const board = Array.from( 
    { length: 8 }, 
    () => Array.from( { length:  8 } )
  );
 
   board[3][4] = new Piece("black");
   board[4][3] = new Piece("black");
   board[3][3] = new Piece("white");
   board[4][4] = new Piece("white");
  
   return board
  // Array.from( { num_of_row_spots } ,  builder_function_for_more_columns )
  // first builds # of rows, second "callback" reads as "for each row, build X columns"
}

/**
 * Constructs a Board with a starting grid set up.
 */
function Board () {
  this.grid = _makeGrid();
}

Board.DIRS = [
  [ 0,  1], [ 1,  1], [ 1,  0],
  [ 1, -1], [ 0, -1], [-1, -1],
  [-1,  0], [-1,  1]
];

/**
 * Returns the piece at a given [x, y] position,
 * throwing an Error if the position is invalid.
 */
Board.prototype.getPiece = function (pos) {
  if (this.isValidPos(pos)) {
    return this.posToBe(pos);
  } else {
    throw Error;
  }
};

/**
 * Checks if there are any valid moves for the given color.
 */
Board.prototype.hasMove = function (color) {
};

/**
 * Checks if the piece at a given position
 * matches a given color.
 */
Board.prototype.isMine = function (pos, color) {
  if(this.getPiece(pos)){
    return(this.posToBe(pos).color === color);
  }
  return false;
};

/**
 * Checks if a given position has a piece on it.
 */
Board.prototype.isOccupied = function (pos) {
   return(!!this.getPiece(pos) );
};


Board.prototype.posToBe = function(pos){
  // console.log(pos)
  return (this.grid[pos[0]][pos[1]]);
}
/**
 * Checks if both the white player and
 * the black player are out of moves.
 */
Board.prototype.isOver = function () {
};

/**
 * Checks if a given position is on the Board.
 */
Board.prototype.isValidPos = function (pos) {
  if ((pos[0] >= 0 && pos[0] < 8) && (pos[1] >= 0 && pos[1] < 8)) {
    return true;
  } else {
    return false;
  }
};

/**
 * Recursively follows a direction away from a starting position, adding each
 * piece of the opposite color until hitting another piece of the current color.
 * It then returns an array of all pieces between the starting position and
 * ending position.
 *
 * Returns null if it reaches the end of the board before finding another piece
 * of the same color.
 *
 * Returns null if it hits an empty position.
 *
 * Returns null if no pieces of the opposite color are found.
 */
function _positionsToFlip (board, pos, color, dir, piecesToFlip=[]) {
  let matchColor = false;
  if(!board.isValidPos(pos)){
    return [];
  }
  if(board.isMine(pos, color) ){
    matchColor = true;
    return [];
  }
  
  let possiblePos = [pos[0] + dir[0], pos[1] + dir[1]];
 
  if ( !board.isValidPos(possiblePos) ) {  // checking if on the board or not, add another statement to check 
    return [];
  } else if ((!board.isMine(possiblePos, color)) && (board.isOccupied(possiblePos))) {
     piecesToFlip.push(possiblePos);
  } else {
    return [];
  }
  
  if(matchColor){
    return piecesToFlip.concat(_positionsToFlip(board, possiblePos, color, dir, piecesToFlip));
  }else{
    return [];
  }

  //if piecesToFlip

  // if the ptf have a length, we can go thru it to check if this move is possible o no
  // we can interate thru the pos we have to combine w/ what we have at start
  // we can check if the 
  // 
  // color at start is same color at the end
}

/**
 * Adds a new piece of the given color to the given position, flipping the
 * color of any pieces that are eligible for flipping.
 *
 * Throws an error if the position represents an invalid move.
 */
Board.prototype.placePiece = function (pos, color) {
};

/**
 * Prints a string representation of the Board to the console.
 */
Board.prototype.print = function () {
  for(let i = 0; i < 8; i++){
    for( let j = 0; j < 8; j++) {
      if (this.grid[i][j]){
        process.stdout.write(`| ${this.grid[i][j].toString()} `)
      }else{
        process.stdout.write(`|${i},${j}`) 
        
      }
    }
    console.log(`|`)
  }


};

/**
 * Checks that a position is not already occupied and that the color
 * taking the position will result in some pieces of the opposite
 * color being flipped.
 */
Board.prototype.validMove = function (pos, color) {
  if (this.isOccupied(pos)) {
    return false;
  }
  
  let positions = [];

  for (let i = 0; i < Board.DIRS.length; i++) {

   positions = positions.concat(_positionsToFlip(this,pos,color,Board.DIRS[i]));
  }
  
 

  if (positions.length > 0) {
    return true;
  } else {
    return false;
  }
};

/**
 * Produces an array of all valid positions on
 * the Board for a given color.
 */
Board.prototype.validMoves = function (color) {
};


module.exports = Board;


let b = new Board()
//b.print()
 console.log(b.validMove([3,4],"black")) // false
 console.log(b.validMove([2,3],"white")) // false
 console.log(b.validMove([2,3],"black")) // true

//  console.log(_positionsToFlip(b,[2,3],"white", [1,0], []))
//  console.log(_positionsToFlip(b,[5,3],"white", [-1,0], []))
 console.log(_positionsToFlip(b,[2,3],"black", [1,0], []))